File: app/src/main/java/com/hover/stax/channels/ChannelDao.java
package com.hover.stax.channels;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;
import androidx.room.Transaction;
import androidx.room.Update;

import com.hover.stax.account.ChannelWithAccounts;

import java.util.List;

@Dao
public interface ChannelDao {

    @Query("SELECT * FROM channels WHERE published = 1 ORDER BY name ASC")
    LiveData<List<Channel>> getAllInAlphaOrder();

    @Query("SELECT * FROM channels WHERE selected = :selected ORDER BY defaultAccount DESC, name ASC")
    LiveData<List<Channel>> getSelected(boolean selected);

    @Query("SELECT COUNT(*) FROM channels WHERE selected = :selected")
    int getSelectedCount(boolean selected);

    @Query("SELECT * FROM channels WHERE id IN (:channel_ids) ORDER BY name ASC")
    LiveData<List<Channel>> getChannels(int[] channel_ids);

    @Query("SELECT * FROM channels WHERE country_alpha2 = :countryCode ORDER BY name ASC")
    List<Channel> getChannels(String countryCode);

    @Query("SELECT * FROM channels WHERE country_alpha2 = :countryCode AND id IN (:channel_ids) ORDER BY name ASC")
    LiveData<List<Channel>> getChannels(String countryCode, int[] channel_ids);

    @Query("SELECT * FROM channels WHERE id = :id LIMIT 1")
    Channel getChannel(int id);

    @Query("SELECT * FROM channels WHERE id = :id LIMIT 1")
    LiveData<Channel> getLiveChannel(int id);

    @Transaction
    @Query("SELECT * FROM channels where selected = :selected ORDER BY name ASC")
    LiveData<List<ChannelWithAccounts>> getChannelsWithAccounts(boolean selected);

    @Query("SELECT COUNT(id) FROM channels")
    int getDataCount();

    @Insert(onConflict = OnConflictStrategy.IGNORE)
    void insertAll(Channel... channels);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insert(Channel channel);

    @Update
    void update(Channel channel);

    @Update
    void updateAll(List<Channel> channel);

    @Delete
    void delete(Channel channel);

    @Query("DELETE FROM channels")
    void deleteAll();
}

File: app/src/main/java/com/hover/stax/channels/ChannelDao.kt
package com.hover.stax.channels

import androidx.lifecycle.LiveData
import androidx.room.*
import com.hover.stax.account.ChannelWithAccounts

@Dao
interface ChannelDao {

    @get:Query("SELECT * FROM channels WHERE published = 1 ORDER BY name ASC")
    val allInAlphaOrder: LiveData<List<Channel>>

    @Query("SELECT * FROM channels WHERE selected = :selected ORDER BY defaultAccount DESC, name ASC")
    fun getSelected(selected: Boolean): LiveData<List<Channel>>

    @Query("SELECT COUNT(*) FROM channels WHERE selected = :selected")
    fun getSelectedCount(selected: Boolean): Int

    @Query("SELECT * FROM channels WHERE id IN (:channel_ids) ORDER BY name ASC")
    fun getChannels(channel_ids: IntArray): LiveData<List<Channel>>

    @Query("SELECT * FROM channels WHERE country_alpha2 = :countryCode ORDER BY name ASC")
    fun getChannels(countryCode: String): List<Channel>

    @Query("SELECT * FROM channels WHERE country_alpha2 = :countryCode AND id IN (:channel_ids) ORDER BY name ASC")
    fun getChannels(countryCode: String, channel_ids: IntArray): LiveData<List<Channel>>

    @Query("SELECT * FROM channels WHERE id = :id LIMIT 1")
    fun getChannel(id: Int): Channel?

    @Query("SELECT * FROM channels WHERE id = :id LIMIT 1")
    fun getLiveChannel(id: Int): LiveData<Channel>

    @Transaction
    @Query("SELECT * FROM channels where selected = 1 ORDER BY name ASC")
    fun getChannelsAndAccounts(): List<ChannelWithAccounts>

    @get:Query("SELECT COUNT(id) FROM channels")
    val dataCount: Int

    @Insert(onConflict = OnConflictStrategy.IGNORE)
    fun insertAll(vararg channels: Channel?)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun insert(channel: Channel?)

    @Update
    fun update(channel: Channel?)

    @Update
    fun updateAll(channel: List<Channel?>?)

    @Delete
    fun delete(channel: Channel?)

    @Query("DELETE FROM channels")
    fun deleteAll()
}